m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -524,58 +519,75 @@ public Thread newThread(Runnable r) {
+                while (!isShutdown.get()) {
+                    try {
+                        // Check whether flush has been requested by the reporter
+                        if (shouldFlush.compareAndSet(true, false)) {
+                            flushBuffered();
+                        }
+
+                        // This `poll` call will block for at most the specified duration to take an item
+                        // off the queue. If we get an item, we append it to the queue to be flushed,
+                        // otherwise we flush what we have so far.
+                        // When this reporter is closed, shutdownNow will be called on the executor,
+                        // which will interrupt this thread and proceed to the `InterruptedException`
+                        // catch block.
+                        SizedMetric sizedMetric = queue.poll(maxBufferingDelay.toMillis(), TimeUnit.MILLISECONDS);
+
+                        if (sizedMetric == null) {
+                            // If we didn't get any new metrics after waiting the specified time,
+                            // flush what we have so far.
+                            flushBuffered();
+                        } else {
+                            process(sizedMetric);
+                        }
+                    } catch (InterruptedException t) {
+                        // no-op
+                    } catch (Throwable t) {
+                        // This is fly-away guard making sure that uncaught exception
+                        // will be logged
+                        LOG.error("Unhandled exception in processor", t);
+                        break;
                     }
-                } catch (InterruptedException t) {
-                    // no-op
-                } catch (Throwable t) {
-                    // This is fly-away guard making sure that uncaught exception
-                    // will be logged
-                    LOG.error("Unhandled exception in processor", t);
-                    break;
                 }
-            }
-
-            state.set(ProcessorState.SHUTDOWN);
-
-            LOG.warn("Processor shutting down");
 
-            shutdown();
-
-            LOG.warn("Processor shut down");
+                shutdown();
+            } finally {
+                // Release shutdown semaphore to notify reporter
+                if (semaphoreAcquired) {
+                    processorsSemaphore.release();
+                }
+            }
         }
 
         private void shutdown() {
-            // Drain queue of any remaining metrics submitted prior to shutdown;
-            runNoThrow(this::drainQueue);
-            // Flush remaining buffers at last (best effort)
-            runNoThrow(this::flushBuffered);
+            LOG.warn("Processor shutting down");
+
+            if (isShutdown.get()) {
+                // If the processor hit an exception, another processor will be created to resume work on the queue,
```

Comment:
Super nit: maybe word this comment to address the conditional it's in? i.e.:

"// Only drain the queue if the entire reporter is shutting down. If the processor hit an exception instead, another processor..."